DART-246 Modify rule S7409: Add Dart language (#5029)
SonarSource/rspec: rules/S7409/dart/rule.adoc

include::../description.adoc[]

include::../ask-yourself.adoc[]

== Recommended Secure Coding Practices

=== Disable JavaScript

If it is possible to disable JavaScript in the WebView, this is the most secure option. 

> When using the https://pub.dev/packages/webview_flutter[`webview_flutter` package] v4 and above, by default 
> JavaScript is disabled in every https://pub.dev/documentation/webview_flutter/latest/webview_flutter/WebViewController-class.html[`WebViewController`]
> instance. Therefore, https://pub.dev/documentation/webview_flutter/latest/webview_flutter/WebViewController/setJavaScriptMode.html[`setJavaScriptMode`]
> does not need to be explicitly called with https://pub.dev/documentation/webview_flutter/latest/webview_flutter/JavaScriptMode.html[`JavaScriptMode.disabled`]. 
> 
> The same applies to https://pub.dev/documentation/webview_flutter/3.0.4/webview_flutter/WebView-class.html[`WebView`] 
> instances in https://pub.dev/packages/webview_flutter[`webview_flutter` package] v3 and below: there is no
> need to explicitly set the named constructor parameter `javascriptMode` to https://pub.dev/documentation/webview_flutter/3.0.4/webview_flutter/JavascriptMode.html[`JavascriptMode.disabled`].

Reviewer: ```suggestion
When using the https://pub.dev/packages/webview_flutter[`webview_flutter` package], JavaScript is disabled by default. Therefore, https://pub.dev/documentation/webview_flutter/latest/webview_flutter/WebViewController/setJavaScriptMode.html[`setJavaScriptMode`] does not need to be explicitly called with https://pub.dev/documentation/webview_flutter/latest/webview_flutter/JavaScriptMode.html[`JavaScriptMode.disabled`] on https://pub.dev/documentation/webview_flutter/latest/webview_flutter/WebViewController-class.html[`WebViewController`] (v4 and above) or https://pub.dev/documentation/webview_flutter/3.0.4/webview_flutter/WebView-class.html[`WebView`] (v3 and below). 
```

This removes repetitions and shortens the paragraphs a bit.

When using the https://pub.dev/packages/flutter_inappwebview[`flutter_inappwebview` package], on the other hand,
the default behavior is to enable JavaScript. Therefore, it is necessary to set the `enableJavaScript` parameter

> to `false` in https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebViewSettings-class.html[InAppWebViewSettings] 
> and https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebViewOptions-class.html[InAppWebViewOptions] 
> instances.

Reviewer: ```suggestion
to `false` in instances of https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebViewSettings-class.html[InAppWebViewSettings] 
and https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/InAppWebViewOptions-class.html[InAppWebViewOptions].
```

> Of course, sometimes it is necessary to enable, or keep enabled, JavaScript, in which case the following 

Reviewer: ```suggestion
Of course, sometimes it is necessary to enable JavaScript, in which case the following 
```

interface when it is no longer needed. If it is needed for a longer time, consider removing it before
loading untrusted content. This can be done by calling `webViewController.removeJavaScriptChannel('channelName')` for the https://pub.dev/packages/webview_flutter[`webview_flutter` package] or `webViewController.removeJavaScriptHandler('channelName')` for the https://pub.dev/packages/flutter_inappwebview[`flutter_inappwebview` package].

A good place to do this in Flutter is inside navigation callbacks like https://pub.dev/documentation/webview_flutter/latest/webview_flutter/NavigationDelegate/onNavigationRequest.html[`NavigationDelegate.onNavigationRequest`] 
for the https://pub.dev/packages/webview_flutter[`webview_flutter` package] or https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/PlatformInAppBrowserEvents/shouldOverrideUrlLoading.html[`WebViewClient.shouldOverrideUrlLoading`] 
for the https://pub.dev/packages/flutter_inappwebview[`flutter_inappwebview` package] package, where you can check the URL being loaded and remove the JavaScript channel 
if the content is untrusted.

=== Alternative methods to implement native bridges

> If a native channel has to be added to the web view or its controller, and it is impossible to remove it 

Reviewer: ```suggestion
If a native channel has to be added to the WebView or its controller, and it is impossible to remove it 
```

combined with https://pub.dev/documentation/flutter_inappwebview/latest/flutter_inappwebview/WebMessageListener-class.html[`WebMessageListener`] 
which allows you to restrict the origins that can send messages to your application.

== Sensitive Code Example

[source,dart]
----
final _controller = WebViewController()
    ..setJavaScriptMode(JavaScriptMode.unrestricted)
    ..addJavaScriptChannel("channel", onMessageReceived: (m) {}); // Sensitive
----

== Compliant Solution

The most secure option is to not enable, or disable, JavaScript entirely. S6362 further explains why it should not 
be enabled unless absolutely necessary.

[source,dart]
----
final _controller = WebViewController();
----

If possible, remove the JavaScript channel after it is no longer needed, or before loading any untrusted content.

[source,dart]
----
_controller.removeJavaScriptChannel("channel");
----

If a JavaScript channel must be used, consider using `runJavascript` (in the case of `flutter_webview`, and 
`evaluateJavascript` in the case of `flutter_inappwebview`) instead. This allows you to restrict the origins that 
can send messages to the JavaScript bridge.

[source,dart]

> ----

Author: The complete example is the following:

```dart
class SecureWebViewExample extends StatefulWidget {
  const SecureWebViewExample({Key? key}) : super(key: key);

  @override
  State<SecureWebViewExample> createState() => _SecureWebViewExampleState();
}

class _SecureWebViewExampleState extends State<SecureWebViewExample> {
  late final WebViewController _controller;
  // List of trusted origins
  final Set<String> _trustedOrigins = {
    'https://example.com',
    'https://yourtrustedsite.com',
  };

  @override
  void initState() {
    super.initState();
    _initWebViewController();
  }

  void _initWebViewController() {
    _controller = WebViewController()
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setNavigationDelegate(
        NavigationDelegate(
          onPageStarted: (String url) {
            debugPrint('Page started loading: $url');
          },
          onPageFinished: (String url) {
            debugPrint('Page finished loading: $url');
            
            // Set up a message receiver in JavaScript that Flutter code can call
            _setupJavaScriptMessageReceiver();
          },
          onNavigationRequest: (NavigationRequest request) {
            // Validate origin before navigation
            final Uri requestUri = Uri.parse(request.url);
            final String origin = '${requestUri.scheme}://${requestUri.host}';
            
            if (_trustedOrigins.contains(origin)) {
              debugPrint('Allowing navigation to: ${request.url}');
              return NavigationDecision.navigate;
            } else {
              debugPrint('Blocking navigation to: ${request.url}');
              return NavigationDecision.prevent;
            }
          },
        ),
      )
      ..loadRequest(Uri.parse('https://example.com'));
  }

  void _setupJavaScriptMessageReceiver() {
    // Inject JavaScript code that sets up a message handler
    _controller.runJavaScript('''
      window.addEventListener('message', function(event) {
        // Origin validation happens on the JavaScript side
        if (${_trustedOrigins.map((origin) => "event.origin === '$origin'").join(' || ')}) {
          // Process message from trusted origin
          window.flutter_inappwebview?.callHandler('messageHandler', 
            JSON.stringify({
              origin: event.origin,
              data: event.data
            })
          );
        } else {
          console.warn('Blocked message from untrusted origin: ' + event.origin);
        }
      }, false);
    ''');
  }

  // Method to safely send messages to the WebView with origin validation
  Future<void> _sendMessageToWebView(String message) async {
    // Get the current URL to validate we're on a trusted page
    final String? currentUrl = await _controller.currentUrl();
    
    if (currentUrl != null) {
      final Uri currentUri = Uri.parse(currentUrl);
      final String origin = '${currentUri.scheme}://${currentUri.host}';
      
      if (_trustedOrigins.contains(origin)) {
        // Safe to send message as we've validated the origin
        await _controller.runJavaScript('''
          // Post a message to the window
          window.postMessage(${jsonEncode(message)}, "${origin}");
        ''');
      } else {
        debugPrint('Blocked sending message to untrusted origin: $origin');
      }
    }
  }

  // Method to receive and process messages from WebView with validation
  Future<void> _processMessageFromWebView() async {
    // Getting the document's origin to validate it's trusted
    final String originScript = 'window.location.origin';
    final String origin = await _controller.runJavaScriptReturningResult(originScript) as String;
    
    if (_trustedOrigins.contains(origin)) {
      // Safe to get and process messages as we've validated the origin
      final String messageScript = 'window.getLatestMessage()'; // Assuming such a function exists
      final String messageJson = await _controller.runJavaScriptReturningResult(messageScript) as String;
      
      // Process the message
      debugPrint('Received valid message from $origin: $messageJson');
      // Process messageJson as needed...
    } else {
      debugPrint('Blocked processing message from untrusted origin: $origin');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Secure WebView Example'),
      ),
      body: WebViewWidget(controller: _controller),
      floatingActionButton: FloatingActionButton(
        onPressed: () => _sendMessageToWebView('Hello from Flutter!'),
        tooltip: 'Send Message',
        child: const Icon(Icons.send),
      ),
    );
  }
}

// Helper function to safely encode JSON
String jsonEncode(dynamic data) {
  // This is a simple implementation - use dart:convert in real code
  return '"${data.toString().replaceAll('"', '\\"')}"';
}
```

The example compiles and run, but I am not sure the origin validation, happing on the JavaScript side, actually makes the approach safer than a channel. WDYT? Should I remove the example?

Reviewer: This honestly requires a proof of concept to test. It goes beyond the specification work in the ticket. Let me discuss with Pierre-Loup tomorrow what to do with this. I do not trust `_processMessageFromWebView` to function correctly with different origins in iframes, but it's hard to say without trying it out in an app.

Reviewer: It may be better to recommend using inappwebview in case a user wants to do this properly.
